File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if not head or not head.next:
            return head

        even_head = head.next

        odd_ptr = head
        even_ptr = head.next
        
        while odd_ptr.next and even_ptr.next:
            odd_ptr.next = even_ptr.next
            even_ptr.next = odd_ptr.next.next
            odd_ptr = odd_ptr.next
            even_ptr = even_ptr.next
        
        odd_ptr.next = even_head

        return head

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode even_head = head.next;

        ListNode odd_ptr = head;
        ListNode even_ptr = head.next;

        while (odd_ptr.next != null && even_ptr.next != null) {
            odd_ptr.next = even_ptr.next;
            odd_ptr = odd_ptr.next;

            even_ptr.next = odd_ptr.next;
            even_ptr = even_ptr.next;
        }

        odd_ptr.next = even_head;

        return head;
    }
}